Mathlib/AlgebraicGeometry/EllipticCurve/Singular/Cusp.lean

```diff
@@ -0,0 +1,122 @@
+/-
+Copyright (c) 2025 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.AlgebraicGeometry.EllipticCurve.Affine.Point
+
+/-!
+
+# Singular cubics with cusps
+
+We develop the basic theory of singular cubics with cusps.
+
+## TODO (Andrew)
+- classify singular cubics with cusps
+
+-/
+
+namespace WeierstrassCurve
+
+variable {R K : Type*} [CommRing R] [Field K] [DecidableEq K]
+
+variable (R) in
+/-- The standard cusp `y² = x³`. -/
+@[simps]
+def standardCusp : WeierstrassCurve R :=
+  ⟨0, 0, 0, 0, 0⟩
+
+@[local simp]
+lemma standardCusp_equation_iff {x y : R} :
+    (standardCusp R).toAffine.Equation x y ↔ y ^ 2 = x ^ 3 := by
+  simp [Affine.equation_iff]
+
+@[local simp]
+lemma standardCusp_nonsingular_iff [IsDomain R] {x y : R} :
+    (standardCusp R).toAffine.Nonsingular x y ↔ y ^ 2 = x ^ 3 ∧ x ≠ 0 ∧ y ≠ 0 := by
+  suffices y ^ 2 = x ^ 3 → ((3 : R) ≠ 0 ∧ x ≠ 0 ∨ (2 : R) ≠ 0 ∧ y ≠ 0 ↔ x ≠ 0 ∧ y ≠ 0) by
+    simpa [Affine.nonsingular_iff, standardCusp_equation_iff, eq_neg_iff_add_eq_zero, ← two_mul]
+  intro H
+  have h : y = 0 ↔ x = 0 := by rw [← pow_eq_zero_iff two_ne_zero, H, pow_eq_zero_iff three_ne_zero]
+  simp only [ne_eq, h, and_self, ← or_and_right, and_iff_right_iff_imp]
+  exact fun _ ↦ not_and_or.mp fun h ↦ one_ne_zero (by linear_combination h.1 - h.2)
+
+/-- The group of nonsigular rational points on the standard cusp `y² = x³`
+is isomorphic (as groups) to `K`. -/
+noncomputable
+def standardCuspEquiv : (standardCusp K).toAffine.Point ≃+ K where
+  toFun
+  | .zero => 0
+  | .some (x := x) (y := y) h => x / y
+  invFun k :=
+    if hk : k = 0 then 0 else
+      .some (x := k⁻¹ ^ 2) (y := k⁻¹ ^ 3) (by simp [pow_right_comm _ 2 3, hk])
+  left_inv
+  | .zero => dif_pos rfl
+  | .some (x := x) (y := y) h => by
+    simp only [div_eq_zero_iff, inv_div]
+    split_ifs with h'
+    · obtain rfl | rfl := h'
+      · by_cases y = 0 <;> simp [Affine.nonsingular_iff', Affine.equation_iff', *] at h
+      · by_cases x = 0 <;> simp [Affine.nonsingular_iff', Affine.equation_iff', *] at h
+    · simp only [Affine.Point.some.injEq]
+      have : y ^ 2 = x ^ 3 := by simpa [Affine.equation_iff', sub_eq_zero] using h.1
+      push_neg at h'
+      constructor
+      · field_simp [this, h'.1]; ring
+      · field_simp [← this, h'.2]; ring
+  right_inv k := by
+    simp only [inv_pow]
+    split_ifs with h'
+    · simp [h']
+    · field_simp [h']; ring
+  map_add'
+  | .zero, .zero => by simp [← Affine.Point.zero_def]
+  | .some h, .zero => by simp [← Affine.Point.zero_def]
+  | .zero, .some h' => by simp [← Affine.Point.zero_def]
+  | .some (x := x₁) (y := y₁) h₁, .some (x := x₂) (y := y₂) h₂ => by
```

Comment:
I thought you could do something like `rintro @⟨x₁, y₁, h₁⟩` etc e.g. see `Affine.Point.toClass`.

Reply:
I think we should just make them explicit. Or else the infoview isn't really readable.